R/mock-session.R

```diff
@@ -407,53 +506,187 @@ MockShinySession <- R6Class(
+    #' @param contentType A string of the content type of the file. Not used by
+    #'   `MockShinySession`.
+    #' @param content A function that takes a single argument file that is a
+    #'   file path (string) of a nonexistent temp file, and writes the content
+    #'   to that file path. (Reactive values and functions may be used from this
+    #'   function.)
+    registerDownload = function(name, filename, contentType, content) {
+      private$file_generators$set(self$ns(name), list(
+        filename = if (is.function(filename)) filename else function() filename,
+        content = content
+      ))
+    },
+    #' @description Get information about the output that is currently being
+    #'   executed.
+    #' @return A list with with the `name` of the output. If no output is
+    #'   currently being executed, this will return `NULL`.
+    getCurrentOutputInfo = function() {
+      name <- private$currentOutputName
+      if (is.null(name)) NULL else list(name = name)
     }
   ),
   private = list(
+    # @field .input Internal ReactiveValues object for normal input sent from client.
     .input = NULL,
+    # @field flushCBs `Callbacks` called before flush.
     flushCBs = NULL,
+    # @field flushedCBs `Callbacks` called after flush.
     flushedCBs = NULL,
+    # @field endedCBs `Callbacks` called when session ends.
     endedCBs = NULL,
+    # @field timer `MockableTimerCallbacks` called at particular times.
     timer = NULL,
-    closed = FALSE,
+    # @field was_closed Set to `TRUE` once the session is closed.
+    was_closed = FALSE,
+    # @field outs List of namespaced output names.
     outs = list(),
+    # @field nsPrefix Prefix with which to namespace inputs and outputs.
     nsPrefix = "mock-session",
+    # @field idCounter Incremented every time `$genId()` is called.
     idCounter = 0,
+    # @field file_generators Map of namespaced output names to lists with
+    #   `filename` and `output` elements, each a function. Updated by
+    #   `$registerDownload()` and read by `$getOutput()`. Files are generated
+    #   on demand when the output is accessed.
+    file_generators = fastmap(),
+    # @field currentOutputName Namespaced name of the currently executing
+    #'   output, or `NULL` if no output is currently executing.
+    currentOutputName = NULL,
+
+    # @description Writes a downloadable file to disk. If the `content`
+    #   function associated with a download handler does not write a file, an
+    #   error is signaled.
+    # @param name The un-namespaced output name associated with the
+    #   downloadable file.
+    # @param download List with two names, `filename` and `content`. Both
+    #   should be functions. `filename` should take no arguments and return a
+    #   string. `content` should accept a path argument and create a file at
+    #   that path.
+    # @return A path to a temp file.
+    renderFile = function(name, download) {
```

Comment:
Should delete all download files when session is closed.

Reply:
Turns out session close callbacks weren't actually implemented and the test for them was broken, so added that, and then this